### ISSUE TITLE
Error: Too many open files (os error 24) for quarto projects

### ISSUE THREAD

**Issue by jmbuhr** [quarto-dev/quarto-cli#5192]
## Setup

```bash
 quarto create project website test --no-open
 cd test
 quarto preview
```

I'm on linux, Pop-OS 22.04, based on Ubuntu 22.04.
I noticed the issue first with Quarto `v1.3.299` (must have appeared between this and early v1.3) , but have also tested with the latest commit of quarto-cli:  024bba6387fc0abe7de1deb2640bd387fc28a692

## Error

```
Preparing to preview
[1/2] about.qmd
[2/2] index.qmd

ERROR: Too many open files (os error 24)

Error: Too many open files (os error 24)
    at new FsWatcher (deno:runtime/js/40_fs_events.js:19:23)
    at Object.watchFs (deno:runtime/js/40_fs_events.js:64:12)
    at runWatcher (file:///home/jannik/projects/quarto-cli/src/project/serve/watch.ts:310:26)
    at Array.forEach (<anonymous>)
    at watchProject (file:///home/jannik/projects/quarto-cli/src/project/serve/watch.ts:351:28)
    at serveProject (file:///home/jannik/projects/quarto-cli/src/project/serve/serve.ts:236:25)
    at async Command.fn (file:///home/jannik/projects/quarto-cli/src/command/preview/cmd.ts:331:9)
    at async Command.execute (file:///home/jannik/projects/quarto-cli/src/vendor/deno.land/x/cliffy@v0.25.4/command/command.ts:1790:7)
    at async quarto (file:///home/jannik/projects/quarto-cli/src/quarto.ts:122:3)
    at async file:///home/jannik/projects/quarto-cli/src/quarto.ts:154:5
```

This only happens for quarto projects, removing `_quarto.yml` allows previewing individual files.

## Related information

Might be related to https://github.com/denoland/deno/issues/17757

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [X] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.
- [X] Please provide the output of `quarto check` so we know which version of quarto and its dependencies you're running.

**Comment by jjallaire** [comment]
I think this limit is often at a user session level (i.e. other processes e.g. IDEs, etc.) count against the limit. It's also possible that there is an unexpected/unwanted interaction with network filesystems?

In any case, the FsWatcher has a lot of problems with both this and with non-standard filesystems (e.g. fuse mounted, nfs, etc.) BTW quarto preview of a document _does not_ use FsWatcher but rather uses polling (I believe in response to an issue you filed a long time ago!).

Needless to say we can't repro this locally. Not sure what is going on in your environment but in any case I think the solution is to move fully away from FsWatcher and create an efficient polling based solution that also works for projects.

**Comment by jmbuhr** [comment]
Interestingly it is not on a network filesystem this time.

However, you are right! After re-logging I was not able to reproduce it until I opened a bunch of different projects that all have language servers running (e.g. quarto-cli with the deno-ls, a neovim configuration with the lua-ls and the quarto-nvim plugin also with the lua-ls), so together they watch a bunch of files.

**Comment by jjallaire** [comment]
You can bump up the user level limit as follows: https://dev.to/rubiin/ubuntu-increase-inotify-watcher-file-watch-limit-kf4

**Comment by bershanskiy** [comment]
There are a few issues at play here: both Deno and Lua use inotify interfaces inefficiently and can exceed the global limit.

If you encounter this issue, you can attempt to work around it using any of these solutions:
 - bump the limit, [as suggested in earlier comment](https://github.com/quarto-dev/quarto-cli/issues/5192#issuecomment-1508605545)
 - make Lua spawn fewer instances, see https://github.com/LuaLS/lua-language-server/issues/1709#issuecomment-1509724678
 - make Deno free spawned instances faster (remove a buggy timeout used to cheat on compat tests), see Fix 1 in https://github.com/darkreader/darkreader/issues/11157
 - make Deno not ignore watch requests (pretend to watch when it does not), see Fix 2 in https://github.com/darkreader/darkreader/issues/11157

**Comment by jjallaire** [comment]
I have implemented polling based preview for website/book projects here: https://github.com/quarto-dev/quarto-cli/commit/9e43c01b74ff16784f0598115a107010c9666518

This will be available once v1.4 pre-release builds start rolling later this week.

**Comment by jjallaire** [comment]
This is not yet available (pending the merge of `feature/new-deno` but will be once that lands.
